src/index.ts

```diff
@@ -0,0 +1,313 @@
+}
+
+
+/**
+ * An interface required for implementing the view-model
+ * for an editor.
+ */
+export
+interface IEditorViewModel {
+  /**
+   * Save the current text buffer to the file.
+   */
+  save(data: string): void;
+  /**
+   * Rename this file.
+   */
+  rename(name: string): void;
+  /**
+   * The mode of the editor, eg. for code this would be
+   * the language, like 'python' or 'javascript'.
+   */
+  mode(): string;
+  /**
+   * Sets the view class on on the view model.
+   */
+  setView(view: IEditorWidget): void;
+  /**
+   * A flag to determine whether to show line numbers.
+   */
+  showLineNumbers: boolean;
+  /**
+   * A flag to determine whether to allow editing.
+   */
+  readOnly: boolean;
+  /**
+   * The number of spaces to insert for each tab.
+   */
+  tabSize: number;
+}
+
+export
+class EditorViewModel implements IDisposable, IEditorViewModel {
+
+  /**
+   * Construct an editor view model.
+   */
+  constructor(model: IEditorModel) {
+    this._model = model;
+  }
+
+  /**
+   * Set the view object for this view model.
+   */
+  setView(view: IEditorWidget): void {
+    this._view = view;
+    EditorWidget
+      .getContentsChanged(this._view)
+      .connect(this._startBufferTimer, this);
+  }
+
+  /**
+   * Save the current buffer in the existing file.
+   */
+  save(): void {
+    this._updateBuffer();
+    this._model.save(this._buffer);
+  }
+
+  /**
+   * Rename the current file.
+   */
+  rename(name: string): void {
+    this._model.rename(name);
+  }
+
+  /**
+   * The mode of the current editor.
+   */
+  mode(): string {
+    return this._mode;
+  }
+
+  /**
+   * Dispose of this object.
+   */
+  dispose(): void {
+    this._disposed = true;
+    clearSignalData(this);
+    this._model = null;
+    this._view = null;
+  }
+
+  /**
+   * Get the disposed status of this object.
+   */
+  get isDisposed(): boolean {
+    return this._disposed;
+  }
+
+  /**
+   * Flag that determines whether to show line numbers.
+   */
+  showLineNumbers = true;
+
+  /**
+   * Flag that determines whether the view is read-only.
+   */
+  readOnly = true;
+
+  /**
+   * Number of spaces to use for a tab.
+   */
+  tabSize = 2;
+
+  private _updateBuffer(): void {
+    if (!this.isDisposed) {
+      var data = this._view.getContents();
+      this._buffer = data;
+    }
+  }
+
+  private _startBufferTimer(): void {
+    clearTimeout(this._bufferTimeoutId);
+    this._bufferTimeoutId = setTimeout(() => {
+      this._updateBuffer();
+    }, this._bufferDelay);
+  }
+
+  private _mode = '';
+  private _disposed = false;
+  private _view: IEditorWidget = null;
+  private _model: IEditorModel;
+  private _buffer = '';
+  /**
+   * The delay in milliseconds between view buffer-fetches.
+   */
+  private _bufferDelay = 1000;
+  /**
+   * The id returned by setTimeout.
+   */
+  private _bufferTimeoutId = 0;
+}
+
+/**
+ * The interface to fulfil in order to implement
+ * an EditorWidget.
+ */
+export
+interface IEditorWidget {
```

Comment:
If you want to make this is an interface (and I don't think we should) it needs to be `interface IEditorWidget extends Widget`